### PR TITLE
[SPARK-34396][SQL] Add a new build-in function delegate

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add a new function `DelegateFunction`.

### Why are the changes needed?

Delegate is just like a big box that can include some other expressions. It will execute all of children and return the last child result as its result.

 

The origin idea is from debug. Debug SQL is hard since SQL is always quite long and complex. This new function can help debug with inject some help functions, e.g., `println, sleep, raise_error`.

 

Two usage examples:
```sql
-- 1
-- raw sql
INSERT INTO TABLE t1
SELECT coalesce(c1, c2) as c FROM t2
 
-- print the column data
INSERT INTO TABLE t1
SELECT delegate(
java_method('scala.Console', 'println', concat('c1: ', c1, ', c2: ', c2)),
coalesce(c1, c2)
) as c FROM t2
 
-- 2
-- raw sql
SELECT if(spark_partition_id() = 1, c1, raise_error('test error')) FROM t2
 
-- add a sleep time before throw error
SELECT if(spark_partition_id() = 1, c1, delegate(
java_method('java.lang.Thread', 'sleep', 3000l),
raise_error('test error')
)) FROM t2
``` 

### Does this PR introduce _any_ user-facing change?

Yes, a new function.

### How was this patch tested?

Add test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
Seems such function is not exists in other database, would you mind to take a look at this idea ? @maropu @cloud-fan @dongjoon-hyun @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 1]
```diff
@@ -269,3 +269,62 @@ case class TypeOf(child: Expression) extends UnaryExpression {
...
+  override def foldable: Boolean = children.forall(_.foldable)
+  override def nullable: Boolean = lastChild.nullable
+  override def dataType: DataType = lastChild.dataType
+
+  override def eval(input: InternalRow): Any = {
+    var result: Any = null
+    children.foreach { child =>
+      result = child.eval(input)
+    }
+    result
```
Comment (AngersZhuuuu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:300]: Hmmm how about add a result map and avoid re-compute same child?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:300]: Not sure what do you mean `same child` ?  This function just execute child one by one.

**Comment by rednaxelafx** [comment]
Just some random thoughts:

This kind of functionality is certain useful for debugging and could come in handy for writing shorthands, but I have mixed feelings about exposing it on the SQL language extension level.
i.e. I'd love to see this if we're designing a new language from scratch, but I'm not so sure about adding it to SQL.

- In terms of the IR (intermediate representation) node, it's very common in programming languages or their compilers to feature this kind of expression type. Some example names are `BlockExpression` (e.g. in [System.Linq.Expressions.BlockExpression](https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.blockexpression)), or C's [Comma expression](https://en.wikipedia.org/wiki/Comma_operator), or just Seq / Sequence expression etc. I haven't seen the term "delegate expression" used this way, though.
- While this comes in really handy, it encourages two things:
  - side effects: yes you do want the side effects for debugging, but for regular SQL queries you should probably avoid side effects to have less surprises
  - duplicate notion of projection: the `DelegateExpression` being proposed here is really just a twp-level nested projection embedded on the expression level: 1 level for evaluating all the children expressions, and another level to discard all but the last expression's result. This duplication is somewhat annoying, as is the case for some of Spark's higher-ordered functions -- they duplicate some semantics from the plan level to the expression, but the optimizations were not updated to recognized them efficiently.
- If we really want to make the most out of the opportunity of adding a new expression infrastructure, we might want to also allow this kind of "block expression" to support "local alias", or more commonly known as just "local variables" or "local bindings" in a programming language. Again, this is something that already exists in the `Project` plan level operator, but it's just clumsy to use in this kind of context.

If we're talking about making some enhancements in the internals of Catalyst and adding a more flexible, more powerful way of representing projections, both on the plan level and on the expression level, that'd be super awesome and I'd help push it forward.
Otherwise adding a one-off expression as proposed in this PR here doesn't feel very clean. At least I don't feel like the proposal as-is should be a regular built-in function, but maybe just something that's on the side for aiding debugging.

Just my two cents

**Comment by ulysses-you** [comment]
thankd you @rednaxelafx  your sharing is always good enough.

My origin requirement is finding a way which behavior is similar with `BlockExpression`, the name `delegate` is just a alias. I see what you worried about, it may make plan and expression fuzzy that's we don't expect.

"local alias" to me more like a udf but we build it in SQL directly, not sure it's accpted since it introduces some complex. Besides, the idea enhancements the Catalyst sounds excited to me, I'd like to do this. Maybe I need sometime to think about it.

**Pull request closed without merging**